api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -199,57 +214,60 @@ public void setResearchPurposeApproved(String ns, String firecloudName, boolean
     saveWithLastModified(workspace, now);
   }
 
-  @Override
-  public Workspace updateUserRoles(Workspace workspace, Set<WorkspaceUserRole> userRoleSet) {
-    Map<Long, WorkspaceUserRole> userRoleMap = new HashMap<Long, WorkspaceUserRole>();
-    for (WorkspaceUserRole userRole : userRoleSet) {
-      userRole.setWorkspace(workspace);
-      userRoleMap.put(userRole.getUser().getUserId(), userRole);
+  private void updateFirecloudAclsOnUser(
+      WorkspaceAccessLevel updatedAccess, WorkspaceACLUpdate currentUpdate) {
+    if (updatedAccess == WorkspaceAccessLevel.OWNER) {
+      currentUpdate.setCanShare(true);
+      currentUpdate.setCanCompute(true);
+      currentUpdate.setAccessLevel(WorkspaceAccessLevel.OWNER.toString());
+    } else if (updatedAccess == WorkspaceAccessLevel.WRITER) {
+      currentUpdate.setCanShare(false);
+      currentUpdate.setCanCompute(true);
+      currentUpdate.setAccessLevel(WorkspaceAccessLevel.WRITER.toString());
+    } else if (updatedAccess == WorkspaceAccessLevel.READER) {
+      currentUpdate.setCanShare(false);
+      currentUpdate.setCanCompute(false);
+      currentUpdate.setAccessLevel(WorkspaceAccessLevel.READER.toString());
+    } else {
+      currentUpdate.setCanShare(false);
+      currentUpdate.setCanCompute(false);
+      currentUpdate.setAccessLevel(WorkspaceAccessLevel.NO_ACCESS.toString());
     }
-    ArrayList<WorkspaceACLUpdate> updateACLRequestList = new ArrayList<WorkspaceACLUpdate>();
-    Iterator<WorkspaceUserRole> dbUserRoles = workspace.getWorkspaceUserRoles().iterator();
-    while (dbUserRoles.hasNext()) {
-      WorkspaceUserRole currentUserRole = dbUserRoles.next();
+  }
+
+  @Override
+  public Workspace updateUserRoles(
+      Workspace workspace, Map<User, WorkspaceAccessLevel> userRoleMap) {
+    // userRoleMap is a map of the new permissions for ALL users on the ws
+    Map<User, WorkspaceAccessEntry> aclsMap = getFirecloudWorkspaceAcls(workspace);
```

Comment:
Using an etag here would be ideal; does FC support it? If not can we file the FR? Either that or we track the delta client-side and only send that (could be a follow-up ticket).